File: Week1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>
#include <stdlib.h>

bool check(long card, int len_card)
{
    bool res = false;
    int n = 0;
    int temp, mas[len_card];
    while (card > 0)
    {
        mas[n] = card % 10;
        //printf("%i\n", mas[n]);
        card = card / 10;
        n++;
    }
    int sum = mas[0];
    for (int x = 1; x < len_card; x++)
    {
        if (x % 2)
        {
            temp = mas[x] * 2;
            //printf("mas[%i] = %i, temp = %i;\n", x, mas[x], temp);
            if (temp > 9)
            {
                temp = trunc(temp / 10) + temp % 10;
            }
        }
        else
        {
            temp = mas[x];
        }
        //printf("%i+", temp);
        sum = sum + temp;
    }
    //printf("\n%i\n", sum);
    if (!(sum % 10))
    {
        res = true;
    }
    else
    {
        printf("INVALID\n");
    }
    return res;
}

int main(void)
{
    long card;
    int len_card;
    //do
    //{
    card = get_long("Number: "); //put card number
    len_card = ceil(log10(card + 1));
    //printf("%i\n", len_card);
    //}
    //while ( (len_card < 13) || (len_card > 16));
    if (len_card < 13 || len_card > 16 || len_card == 14)
    {
        printf("INVALID\n");
        exit(0);
    }
    int emitent = trunc(card / pow(10, len_card - 2));
    if (trunc(emitent / 10) == 4)
    {
        if (check(card, len_card))
        {
            printf("VISA\n");
        }
    }
    else if (emitent > 50 && emitent < 56)
    {
        if (check(card, len_card))
        {
            printf("MASTERCARD\n");
        }
    }
    else if (emitent == 34 || emitent == 37)
    {
        if (check(card, len_card))
        {
            printf("AMEX\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
    //printf("%i\n", emitent);
}

File: Week6/sentimental-credit/credit.py
def main():
    while True:
        card = input("Number: ")
        if card.isdigit():
            break
        else:
            print("Please put a card number (only digits)")  # if card contain not only digits
    if 13 < len(card) < 16 or len(card) != 14:
        if is_visa(card):
            print("VISA")
            exit()
        elif is_master(card):
            print("MASTERCARD")
            exit()
        elif is_amex(card):
            print("AMEX")
            exit()
    print("INVALID")


def is_visa(card):  # Check if is VISA card
    return check(card) if int(card[:1]) == 4 else False


def is_master(card):  # Check if is MASTERCARD card
    return check(card) if 50 < int(card[:2]) < 56 else False


def is_amex(card):  # Check if is AMEX card
    return check(card) if int(card[:2]) == 34 or int(card[:2]) == 37 else False


def check(card):  # Check if the checksum of card is right
    check_sum = sum(int(item) for item in card[::-2])
    for i in card[-2::-2]:
        i = int(i)
        if i * 2 >= 10:
            check_sum += (i * 2 // 10) + (i * 2 % 10)
        else:
            check_sum += i * 2
    if not check_sum % 10:
        return True
    return False


if __name__ == "__main__":
    main()
